### PR TITLE
feat: Opensearch construct

### PR DESCRIPTION
**Pull request by alexvt-amz**
**Issue #, if available:**
#203 

## Description of changes:

Opensearch construct

**Checklist**

 

* [X] Update tests
* [X] Update docs
* [X] PR title follows [conventional commit semantics](https://www.conventionalcommits.org/en/v1.0.0/) (`fix: `, `feat: `, `docs: `, ...)

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by lmouhib — CHANGES_REQUESTED** [review]
Can you please move the construct under a new directory called `consumption`

**Comment by alexvt-amz** [comment]
> Can you please move the construct under a new directory called `consumption`

done

**Comment by sharp-pixel** [comment]
Can we rename `Opensearch` to `OpenSearch`?

**Comment by lmouhib** [comment]
> Can we rename `Opensearch` to `OpenSearch`?

@sharp-pixel where ? in the documentation? or Construct name? or everywhere where the uppercase is compliant?

**Comment by sharp-pixel** [comment]
> > Can we rename `Opensearch` to `OpenSearch`?
> 
> @sharp-pixel where ? in the documentation? or Construct name? or everywhere where the uppercase is compliant?

Everywhere where it doesn't break something.
Examples:
- names: `OpensearchNodes` -> `OpenSearchNodes`
- comments: Default Node Instances for Opensearch cluster -> Default Node Instances for OpenSearch cluster

**Comment by alexvt-amz** [comment]
> > > Can we rename `Opensearch` to `OpenSearch`?
> > 
> > 
> > @sharp-pixel where ? in the documentation? or Construct name? or everywhere where the uppercase is compliant?
> 
> Everywhere where it doesn't break something. Examples:
> 
> * names: `OpensearchNodes` -> `OpenSearchNodes`
> * comments: Default Node Instances for Opensearch cluster -> Default Node Instances for OpenSearch cluster

Fixed

File: framework/src/consumption/lib/opensearch/opensearch-props.ts [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+  readonly masterNodeInstanceType?: string;
+  readonly masterNodeInstanceCount?: number;
+  readonly warmInstanceType?:number;
+  readonly warmInstanceCount?: number;
+  readonly multiAzWithStandbyEnabled?: boolean;
+  readonly ebsVolumeType?: EbsDeviceVolumeType;
+  readonly ebsSize?: number;
+  readonly samlEntityId:string;
+  readonly samlMetadataContent:string;
+  readonly samlMasterBackendRole:string;
```
Comment (vgkowski) [inline framework/src/consumption/lib/opensearch/opensearch-props.ts:51]: strong type => IRole or Role

File: framework/src/consumption/lib/opensearch/opensearch-props.ts [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+  readonly samlRolesKey?:string;
+  readonly samlSubjectKey?:string;
+  readonly samlSessionTimeoutMinutes?:number;
+  readonly enableAutoSoftwareUpdate?: boolean;
+  readonly enableVersionUpgrade?: boolean;
+  readonly encryptionKmsKeyArn?:string;
```
Comment (vgkowski) [inline framework/src/consumption/lib/opensearch/opensearch-props.ts:57]: strong type => IKey or Key

File: framework/src/consumption/lib/opensearch/opensearch.ts [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+
+/**
+ * A construct to provision Amazon Openssearch Cluster and OpenSearch Dashboards.
+ * Uses IAM Identity Center SAML authentication.
+ * If OpenSearch cluster is deployed in vpc created using DataVpc construct,
+ * ClientVPNEndpoint will be provisioned automatically for secure access to Opnesearch Dashboards.
+ *
+ * @example
+ *
+ *    const osCluster = new dsf.consumption.OpensearchCluster(this, 'MyOpensearchCluster',{
```
Comment (vgkowski) [inline framework/src/consumption/lib/opensearch/opensearch.ts:24]: reindent on the left

File: framework/src/consumption/lib/opensearch/opensearch.ts [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+ *    osCluster.addRoleMapping('dashboards_user','<IAMIdentityCenterDashboardUsersGroupId>');
+ *    osCluster.addRoleMapping('readall','<IAMIdentityCenterDashboardUsersGroupId>');
+ *
+ *
+*/
+
+export class OpensearchCluster extends TrackedConstruct {
+
+  /**
+   * @public
```
Comment (vgkowski) [inline framework/src/consumption/lib/opensearch/opensearch.ts:42]: @public unnecessary

File: framework/src/consumption/lib/opensearch/opensearch.ts [hunk 5]
```diff
@@ -0,0 +1,305 @@
...
+   * OpenSearchCluster domain
+   */
+  public readonly domain: Domain;
+
+  /**
+   * @public
+   * Cloudwatch log group to store OpenSearch cluster logs
```
Comment (vgkowski) [inline framework/src/consumption/lib/opensearch/opensearch.ts:49]: => CloudWatch Logs Log Group

File: framework/src/consumption/lib/opensearch/opensearch.ts [hunk 6]
```diff
@@ -0,0 +1,305 @@
...
+ *    osCluster.addRoleMapping('readall','<IAMIdentityCenterDashboardUsersGroupId>');
+ *
+ *
+*/
+
+export class OpensearchCluster extends TrackedConstruct {
+
+  /**
+   * @public
+   * OpenSearchCluster domain
```
Comment (vgkowski) [inline framework/src/consumption/lib/opensearch/opensearch.ts:43]: => OpenSearch Cluster Domain

File: framework/src/consumption/lib/opensearch/opensearch.ts [hunk 7]
```diff
@@ -0,0 +1,305 @@
...
+  /**
+   * Constructs a new instance of the OpenSearchCluster class
+   * @param {Construct} scope the Scope of the AWS CDK Construct
+   * @param {string} id the ID of the AWS CDK Construct
+   * @param {OpensearchClusterProps} props the OpenSearchCluster [properties]{@link OpensearchClusterProps}
+   */
+
+  constructor(scope: Construct, id: string, props: OpensearchProps) {
+    const trackedConstructProps: TrackedConstructProps = {
+      trackingTag: id,
```
Comment (vgkowski) [inline framework/src/consumption/lib/opensearch/opensearch.ts:89]: => trackingTag: OpenSearchCluster.name

File: framework/test/unit/consumption/opensearch.test.ts [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+  // Instantiate AccessLogsBucket Construct with default
+  new OpensearchCluster(stack, 'OpensearchVpc', {
+    domainName: 'mycluster2',
+    samlEntityId: '<idpTest>',
+    samlMetadataContent: 'xmlContent',
+    samlMasterBackendRole: 'IdpGroupId',
+    deployInVpc: true,
+    removalPolicy: RemovalPolicy.DESTROY,
+  });
+
```
Comment (vgkowski) [inline framework/test/unit/consumption/opensearch.test.ts:32]: Missing tests for extra methods based on custom resource

**Review by vgkowski — CHANGES_REQUESTED** [review]

File: framework/src/consumption/lib/opensearch/opensearch-props.ts [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+  readonly deployInVpc:boolean;
+  readonly vpc?: IVpc;
```
Comment (vgkowski) [inline framework/src/consumption/lib/opensearch/opensearch-props.ts:59]: Subnets are missing

**Review by vgkowski — CHANGES_REQUESTED** [review]

**Review by vgkowski — APPROVED** [review]

**Review by lmouhib — APPROVED** [review]

**Pull request merged**